modules/db/src/main/scala/gem/dao/GcalDao.scala

```diff
@@ -10,48 +10,62 @@ import java.time.Duration
 
 object GcalDao {
 
-  def insert(gcal: GcalConfig, step: Option[Int]): ConnectionIO[Int] = {
-    val arcs: GcalArc => Boolean = gcal.arcs.member
-
-    for {
-      _ <- sql"""INSERT INTO gcal (step_id, continuum, ar_arc, cuar_arc, thar_arc, xe_arc, filter, diffuser, shutter, exposure_time, coadds)
-                      VALUES ($step, ${gcal.continuum}, ${arcs(ArArc)}, ${arcs(CuArArc)}, ${arcs(ThArArc)}, ${arcs(XeArc)}, ${gcal.filter}, ${gcal.diffuser}, ${gcal.shutter}, ${gcal.exposureTime}, ${gcal.coadds})
-              """.update.run
-      id <- sql"select lastval()".query[Int].unique
-    } yield id
-  }
+  def insert(gcal: GcalConfig, step: Option[Int]): ConnectionIO[Int] =
+    Statements.insert(gcal, step).withUniqueGeneratedKeys[Int]("gcal_id") // janky, hm
+
+  def select(id: Int): ConnectionIO[Option[GcalConfig]] =
+    Statements.select(id).option.map(_.flatten)
+
+  object Statements {
 
-  private case class GcalKernel(
-    continuum: Option[GcalContinuum],
-    ar_arc:    Boolean,
-    cuar_arc:  Boolean,
-    thar_arc:  Boolean,
-    xe_arc:    Boolean,
-    filter:    GcalFilter,
-    diffuser:  GcalDiffuser,
-    shutter:   GcalShutter,
-    expTime:   Duration,
-    coadds:    Int)
-  {
-    def toGcalConfig: Option[GcalConfig] =
-      GcalLamp.fromConfig(continuum, ArArc -> ar_arc, CuArArc -> cuar_arc, ThArArc -> thar_arc, XeArc -> xe_arc).map { lamp =>
-        GcalConfig(lamp, filter, diffuser, shutter, expTime, coadds)
+    // CoAdds has a DISTINCT type due to its check constraint so we need a fine-grained mapping
```

Comment:
It's a different type from JDBC's point of view, so we need to tell doobie that it's not an error. That's really all that's going on here. I did it case by case but may be able to abstract some stuff out.